### PR TITLE
Delay registering of observables until Observable.ready

### PR DESCRIPTION
**Pull request by andypetrella**
Delay registering of observables until `Observable.ready` event.

This also fixes issues with lazy kernel load.

### CODE REVIEW

File: public/javascripts/notebook/observable.js [hunk 1]
```diff
@@ -113,31 +113,47 @@ return new function () {
 
 
 
-  this.makeObservableHelper = function (id, kind, initialValue) {
+  this.makeObservableHelper = function (id, kind, initialValue, df) {
+    var _df = df
+    if (_.isUndefined(_df)) {
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: `_` is for `underscore` ? not included/required in `define` header.

Comment (andypetrella) [inline public/javascripts/notebook/observable.js]: damn good catch

File: public/javascripts/notebook/observable.js [hunk 2]
```diff
@@ -113,31 +113,47 @@ return new function () {
...
     if(!this.isInitialized()) {
       // in case the Observable isn't initialized, we differ the creation of the observable
       var args = Array.prototype.slice.call(arguments);
       var me = this;
+      var current_df = _df;
       console.warn("Delaying the creation of Observable (" + args.join(",") + ") since Observable isn't initialized yet");
       console.log("Observable not initialized:", me);
       events.on('Observable.ready', function() {
-        me.makeObservableHelper.apply(me, args);
+        me.makeObservableHelper.apply(me, args, current_df);
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: hmm does `.apply()` take 3 args? so it should be `apply(this, args)`.
thus we just need to replace the last argument this the deferred `current_df`.

File: public/javascripts/notebook/observable.js (hunk 1 above)
Comment (andypetrella) [inline public/javascripts/notebook/observable.js]: done

File: public/javascripts/notebook/observable.js (hunk 2 above)
Comment (andypetrella) [inline public/javascripts/notebook/observable.js]: woopsy.... thanks lovely js and its dynamic number of unspecified args...

File: public/javascripts/notebook/observable.js [hunk 3]
```diff
@@ -113,31 +113,47 @@ return new function () {
...
-      observable = ko[kind](initialValue);
-      this.register_observable(id, observable);
-    } else if (typeof initialValue !== 'undefined') {
-      this.observableSetIfChanged(observable, initialValue);
     } else {
-      console.error("Cannot register observable with id '" + id + "', kind '" + kind + "'" + "', initialValue '" + initialValue + "'")
-      console.log("Observable is ", this);
-      console.log("Registered observables are ", this.observables);
+      var observable = this.observables[id];
+      if (typeof observable === 'undefined') {
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: use `_.isUndefined` ?

File: public/javascripts/notebook/observable.js [hunk 4]
```diff
@@ -113,31 +113,48 @@ return new function () {
...
+        this.df.done(function(ob) {
+          ob.subscribe(subscriber);
+        });
+      };
+    }
     if(!this.isInitialized()) {
       // in case the Observable isn't initialized, we differ the creation of the observable
       var args = Array.prototype.slice.call(arguments);
       var me = this;
+      args.push(_df);
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: this is a bit fishy.
1. we pass less arguments than required when calling this fn: `return this.makeObservableHelper(id, 'observable', initialValue)`
2. not quite sure what happens in `Array.prototype.slice.call(arguments);`, but doing only `args.push(_df)` isn't that we always assume to receive less arguments than required ?

File: public/javascripts/notebook/observable.js [hunk 5]
```diff
@@ -113,31 +113,48 @@ return new function () {
...
       console.warn("Delaying the creation of Observable (" + args.join(",") + ") since Observable isn't initialized yet");
       console.log("Observable not initialized:", me);
       events.on('Observable.ready', function() {
+        args
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: what's this?

File: public/javascripts/notebook/observable.js [hunk 6]
```diff
@@ -113,31 +113,48 @@ return new function () {
 
 
 
-  this.makeObservableHelper = function (id, kind, initialValue) {
+  this.makeObservableHelper = function (id, kind, initialValue, df) {
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: this is rather complex now. maybe add a comment of what's happening here, something like
- until observable is not initialized we create a magic fn which:
  * simulates a delayed observable  by delaying `.subscribe` calls via a `$.Deferred`
  * passes over the `publishValue` to observable (?) once it's available
- .... (something else about lazy observable initialization)

File: public/javascripts/notebook/observable.js [hunk 7]
```diff
@@ -113,31 +113,48 @@ return new function () {
...
+        this.df.done(function(ob) {
+          ob(publishValue);
+        });
+      }
+      _df.df = $.Deferred();
+      _df.subscribe = function(subscriber) {
+        this.df.done(function(ob) {
+          ob.subscribe(subscriber);
+        });
+      };
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: btw: `Uncaught TypeError: Observable.makeObservable(...).extend is not a function`

File: public/javascripts/notebook/observable.js [hunk 8]
```diff
@@ -1,5 +1,29 @@
-define(['jquery', 'base/js/events', 'knockout', 'equiv'], function ($, events, ko, equiv) {
+define(['jquery', 'base/js/events', 'knockout', 'equiv', 'underscore'], function ($, events, ko, equiv, _) {
+ko.extenders.async = function(computedDeferred, initialValue) {
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: looks like this is based on:
https://github.com/knockout/knockout/wiki/asynchronous-dependent-observables
or https://smellegantcode.wordpress.com/2012/12/10/asynchronous-computed-observables-in-knockout-js/
add a link and explanation maybe?

Comment (andypetrella) [inline public/javascripts/notebook/observable.js]: damn, need to be deleted...

Comment (vidma) [inline public/javascripts/notebook/observable.js]: > need to be deleted... 
?

**Comment by vidma** [comment]
hmm, from what I tried a little (need more testing), seems to work like a charm (both in lazy kernel and eager kernel modes)!

didn't check the new code carefully yet...

File: public/javascripts/notebook/observable.js (hunk 8 above)
Comment (andypetrella) [inline public/javascripts/notebook/observable.js]: yeah not needed

**Comment by vidma** [comment]
certainly tomorrow I'm deploying this to our cluster ;)

File: public/javascripts/notebook/observable.js [hunk 9]
```diff
@@ -111,25 +111,27 @@ return new function () {
...
-      var me = this;
-      console.warn("Delaying the creation of Observable (" + args.join(",") + ") since Observable isn't initialized yet");
-      console.log("Observable not initialized:", me);
-      events.on('Observable.ready', function() {
-        me.makeObservableHelper.apply(me, args);
-      });
-      return;
-    }
     var observable = this.observables[id];
     if (typeof observable === 'undefined') {
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: use `_.isUndefined` ? ok to refactor in separate PR too.

File: public/javascripts/notebook/observable.js [hunk 10]
```diff
@@ -111,25 +111,27 @@ return new function () {
...
     	console.log("Creating new observable (client request): " + id)
       observable = ko[kind](initialValue);
-      this.register_observable(id, observable);
+
+      if(!this.isInitialized()) {
+        // in case the Observable isn't initialized, we differ the creation of the observable
+        var me = this;
+        var lid = id;
+        var ob = observable;
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: do we need `lid` & `ob` variables ?

File: public/javascripts/notebook/observable.js [hunk 11]
```diff
@@ -111,25 +111,27 @@ return new function () {
...
-      return;
-    }
     var observable = this.observables[id];
     if (typeof observable === 'undefined') {
     	console.log("Creating new observable (client request): " + id)
       observable = ko[kind](initialValue);
-      this.register_observable(id, observable);
+
+      if(!this.isInitialized()) {
+        // in case the Observable isn't initialized, we differ the creation of the observable
```
Comment (vidma) [inline public/javascripts/notebook/observable.js]: I'd rephrase this comment: delay registering of knockout observable until `Observable.ready`

**Comment by vidma** [comment]
looks :ok:  now. :clap: 

see the few minor comments.
and I think you want to squash the commits into one before merging ;)

**Comment by andypetrella** [comment]
LGTM!
